FFM I2C  SMBus protocol does not support reading multiregister devices.

Regression testing with a AT24C512 EEPROM.     Test attempts to read data pointed at by  a two byte register address.  
[main] TRACE com.pi4j.devices.at24c512.AT24C512 - >>> enter: readEEPROM register 0342  numByte : 2
Exception in thread "main" java.lang.UnsupportedOperationException: SMBus protocol does not support reading multiregister devices. Please, use I2CDirect or I2CFile provider instead.
	at com.pi4j.plugin.ffm@4.0.0-SNAPSHOT/com.pi4j.plugin.ffm.providers.i2c.impl.I2CSMBus.readRegister(I2CSMBus.java:121)
	at com.pi4j@4.0.0-SNAPSHOT/com.pi4j.io.i2c.I2CRegisterDataReader.readRegister(I2CRegisterDataReader.java:139)
	at com.pi4j@4.0.0-SNAPSHOT/com.pi4j.io.i2c.I2CRegisterDataReader.readRegister(I2CRegisterDataReader.java:171)
	at com.pi4j.devices/com.pi4j.devices.at24c512.AT24C512.readEEPROM(AT24C512.java:228)
	at com.pi4j.devices/com.pi4j.devices.at24c512.AT24C512_App.main(AT24C512_App.java:208)
     


This is a device that requires a REPEATED RESTART.  I will include info from linuxfs and the datasheet
.  
Random Read
A random read begins in the same way as a byte write operation does to load in a new data word address. This is
known as a “dummy write” sequence; however, the data byte and the Stop condition of the byte write must be omitted
to prevent the part from entering an internal write cycle. Once the device address and word address are clocked in
and acknowledged by the EEPROM, the bus host must generate another Start condition. The bus host now initiates
a current address read by sending a Start condition, followed by a valid device address byte with the R/W bit set to
logic ‘1’. The EEPROM will ACK the device address and serially clock out the data word on the SDA line. All types
of read operations will be terminated if the bus host does not respond with an ACK (it NACKs) during the ninth clock
cycle. After the NACK response, the host may send a Stop condition to complete the protocol, or it can send a Start
condition to begin the next sequence.



<img width="824" height="257" alt="Image" src="https://github.com/user-attachments/assets/9fb2503c-4a77-4219-a896-ae7e3e1d0a3e" />

<img width="824" height="257" alt="Image" src="https://github.com/user-attachments/assets/baf28542-3e67-4a5e-8c45-35087fb0fdae" />

Using I2CImplementation.FILE

` var i2cDeviceConfig = I2C.newConfigBuilder(this.pi4j)
            .bus(bus)
            .device(address)
            .name(name)
            .i2cImplementation(I2CImplementation.FILE)
            .provider("ffm-i2c")  /* ffm-i2c  "FFMI2CProviderImpl" "linuxfs-i2c")  "linuxfs-i2c") */
            .build();
        this.config = i2cDeviceConfig;
        this.i2c = this.pi4j.create(i2cDeviceConfig);
        
[main] TRACE com.pi4j.devices.at24c512.AT24C512 - <<< exit: i2cDetail  com.pi4j.plugin.ffm.providers.i2c.impl.I2CFile@a74868d bus : 1  address : 80
[main] INFO com.pi4j.util.Console -   Dev I2C detail    com.pi4j.plugin.ffm.providers.i2c.impl.I2CFile@a74868d bus : 1  address : 80
[main] INFO com.pi4j.util.Console -   Setup ----------------------------------------------------------
[main] TRACE com.pi4j.devices.at24c512.AT24C512 - >>> enter: i2cDetail
[main] TRACE com.pi4j.devices.at24c512.AT24C512 - <<< exit: i2cDetail  com.pi4j.plugin.ffm.providers.i2c.impl.I2CFile@a74868d bus : 1  address : 80
[main] INFO com.pi4j.util.Console -   I2C detail : com.pi4j.plugin.ffm.providers.i2c.impl.I2CFile@a74868d bus : 1  address : 80
[main] TRACE com.pi4j.devices.at24c512.AT24C512 - >>> enter: writeEEPROM register :0342  numBytes 2
Exception in thread "main" com.pi4j.exception.Pi4JException: Failed to execute action for device 80 on bus 1
	at com.pi4j@4.0.0-SNAPSHOT/com.pi4j.io.i2c.I2CBusBase._execute(I2CBusBase.java:57)
	at com.pi4j.plugin.ffm@4.0.0-SNAPSHOT/com.pi4j.plugin.ffm.providers.i2c.I2CBusFFM.execute(I2CBusFFM.java:143)
	at com.pi4j.plugin.ffm@4.0.0-SNAPSHOT/com.pi4j.plugin.ffm.providers.i2c.impl.I2CFile.writeRegister(I2CFile.java:101)
	at com.pi4j@4.0.0-SNAPSHOT/com.pi4j.io.i2c.I2CRegisterDataWriter.writeRegister(I2CRegisterDataWriter.java:144)
	at com.pi4j.devices/com.pi4j.devices.at24c512.AT24C512.writeEEPROM(AT24C512.java:256)
	at com.pi4j.devices/com.pi4j.devices.at24c512.AT24C512_App.main(AT24C512_App.java:205)
Caused by: com.pi4j.exception.Pi4JException: Failed to execute action for device 80 on bus /dev/i2c-1
	at com.pi4j.plugin.ffm@4.0.0-SNAPSHOT/com.pi4j.plugin.ffm.providers.i2c.I2CBusFFM.lambda$execute$0(I2CBusFFM.java:147)
	at com.pi4j@4.0.0-SNAPSHOT/com.pi4j.io.i2c.I2CBusBase._execute(I2CBusBase.java:44)
	... 5 more
Caused by: com.pi4j.exception.Pi4JException: Error during call to method 'write' with data '[20, [3, 66, 18, 52]]': Bad file descriptor (9)
	at com.pi4j.plugin.ffm@4.0.0-SNAPSHOT/com.pi4j.plugin.ffm.common.file.FileDescriptorNative.write(FileDescriptorNative.java:90)
	at com.pi4j.plugin.ffm@4.0.0-SNAPSHOT/com.pi4j.plugin.ffm.providers.i2c.impl.I2CFile.lambda$writeRegister$2(I2CFile.java:101)
	at com.pi4j.plugin.ffm@4.0.0-SNAPSHOT/com.pi4j.plugin.ffm.providers.i2c.I2CBusFFM.lambda$execute$0(I2CBusFFM.java:145)
	... 6 more
Caused by: com.pi4j.exception.Pi4JException: Error during call to method 'write' with data '[20, [3, 66, 18, 52]]': Bad file descriptor (9)
	at com.pi4j.plugin.ffm@4.0.0-SNAPSHOT/com.pi4j.plugin.ffm.common.Pi4JNativeContext.processError(Pi4JNativeContext.java:56)
	at com.pi4j.plugin.ffm@4.0.0-SNAPSHOT/com.pi4j.plugin.ffm.common.file.FileDescriptorNative.write(FileDescriptorNative.java:87)
	... 8 more

Process finished with exit code 1

`

--- a/plugins/pi4j-plugin-ffm/src/main/java/com/pi4j/plugin/ffm/common/i2c/rdwr/I2CMessage.java
+++ b/plugins/pi4j-plugin-ffm/src/main/java/com/pi4j/plugin/ffm/common/i2c/rdwr/I2CMessage.java
@@ -2,7 +2,6 @@ package com.pi4j.plugin.ffm.common.i2c.rdwr;
 
 import com.pi4j.plugin.ffm.common.Pi4JLayout;
 
-import java.lang.foreign.Arena;
 import java.lang.foreign.MemoryLayout;
 import java.lang.foreign.MemorySegment;
 import java.lang.foreign.SegmentAllocator;
@@ -80,8 +79,6 @@ public record I2CMessage(int address, int flags, int len, byte[] buf) implements
     private static final VarHandle VH_LEN = LAYOUT.varHandle(MemoryLayout.PathElement.groupElement("len"));
     private static final VarHandle VH_BUFFER = LAYOUT.varHandle(MemoryLayout.PathElement.groupElement("buf"));
 
-    private static final SegmentAllocator SEGMENT_ALLOCATOR = Arena.ofAuto();
-
     /**
      * Creates empty I2CMessage object.
      *
@@ -103,21 +100,28 @@ public record I2CMessage(int address, int flags, int len, byte[] buf) implements
         var flags = (int) VH_FLAGS.get(buffer, 0L);
         var len = (int) VH_LEN.get(buffer, 0L);
 
-        var bufferSegment = (MemorySegment)VH_BUFFER.get(buffer, 0L);
-        var bytes = new byte[len];
-        bufferSegment.reinterpret(len).asByteBuffer().get(bytes, 0, len);
+        var bufferAddress = (MemorySegment) VH_BUFFER.get(buffer, 0L);
+        var bufferSegment = bufferAddress.reinterpret(len);
+        var buf = new byte[len];
+        for (int i = 0; i < buf.length; i++) {
+            buf[i] = bufferSegment.getAtIndex(ValueLayout.JAVA_BYTE, i);
+        }
 
-        return new I2CMessage(address, flags, len, bytes);
+        return new I2CMessage(address, flags, len, buf);
     }
 
     @Override
     public void to(MemorySegment buffer) throws Throwable {
+        throw new UnsupportedOperationException("I2CMessage needs to be called with external Segment Allocator");
+    }
+
+    @Override
+    public void to(MemorySegment buffer, SegmentAllocator allocator) throws Throwable {
         VH_ADDRESS.set(buffer, 0L, (short) address);
         VH_FLAGS.set(buffer, 0L, (short) flags);
         VH_LEN.set(buffer, 0L, (short) len);
 
-        var bufferSegment = SEGMENT_ALLOCATOR.allocate(buf.length);
-        bufferSegment.asByteBuffer().put(buf);
+        var bufferSegment = allocator.allocateFrom(ValueLayout.JAVA_BYTE, buf);
         VH_BUFFER.set(buffer, 0L, bufferSegment);
     }
 

--- a/plugins/pi4j-plugin-ffm/src/main/java/com/pi4j/plugin/ffm/common/i2c/rdwr/RDWRData.java
+++ b/plugins/pi4j-plugin-ffm/src/main/java/com/pi4j/plugin/ffm/common/i2c/rdwr/RDWRData.java
@@ -2,7 +2,6 @@ package com.pi4j.plugin.ffm.common.i2c.rdwr;
 
 import com.pi4j.plugin.ffm.common.Pi4JLayout;
 
-import java.lang.foreign.Arena;
 import java.lang.foreign.MemoryLayout;
 import java.lang.foreign.MemorySegment;
 import java.lang.foreign.SegmentAllocator;
@@ -23,8 +22,6 @@ public record RDWRData(I2CMessage[] msgs, int nmsgs) implements Pi4JLayout {
 
     private static final VarHandle VH_MSGS = LAYOUT.varHandle(MemoryLayout.PathElement.groupElement("msgs"));
     private static final VarHandle VH_NMSGS = LAYOUT.varHandle(MemoryLayout.PathElement.groupElement("nmsgs"));
-
-    private static final SegmentAllocator SEGMENT_ALLOCATOR = Arena.ofAuto();
 
     @Override
     public MemoryLayout getMemoryLayout() {
@@ -52,12 +49,17 @@ public record RDWRData(I2CMessage[] msgs, int nmsgs) implements Pi4JLayout {
 
     @Override
     public void to(MemorySegment buffer) throws Throwable {
+        throw new UnsupportedOperationException("RDWRDate needs to be called with external Segment Allocator");
+    }
+
+    @Override
+    public void to(MemorySegment buffer, SegmentAllocator allocator) throws Throwable {
         VH_NMSGS.set(buffer, 0L, nmsgs);
 
-        var bodySegment = SEGMENT_ALLOCATOR.allocate(msgs.length * I2CMessage.LAYOUT.byteSize());
+        var bodySegment = allocator.allocate(msgs.length * I2CMessage.LAYOUT.byteSize());
         for (int i = 0; i < msgs.length; i++) {
             var messageSlice = bodySegment.asSlice(i * I2CMessage.LAYOUT.byteSize(), I2CMessage.LAYOUT.byteSize());
-            msgs[i].to(messageSlice);
+            msgs[i].to(messageSlice, allocator);
             messageSlice.copyFrom(messageSlice);
         }
 

--- a/plugins/pi4j-plugin-ffm/src/main/java/com/pi4j/plugin/ffm/providers/i2c/I2CBusFFM.java
+++ b/plugins/pi4j-plugin-ffm/src/main/java/com/pi4j/plugin/ffm/providers/i2c/I2CBusFFM.java
@@ -86,7 +86,7 @@ public class I2CBusFFM extends I2CBusBase {
     public boolean supportsSMBus() {
         return hasFunctionality(I2CFunctionality.I2C_FUNC_SMBUS_QUICK) || hasFunctionality(I2CFunctionality.I2C_FUNC_SMBUS_BYTE)
             || hasFunctionality(I2CFunctionality.I2C_FUNC_SMBUS_BYTE_DATA) || hasFunctionality(I2CFunctionality.I2C_FUNC_SMBUS_WORD_DATA)
-            || hasFunctionality(I2CFunctionality.I2C_FUNC_SMBUS_I2C_BLOCK);
+            || hasFunctionality(I2CFunctionality.I2C_FUNC_SMBUS_BLOCK_DATA);
     }
 
     public boolean hasFunctionality(I2CFunctionality functionality) {

--- a/plugins/pi4j-plugin-ffm/src/main/java/com/pi4j/plugin/ffm/providers/i2c/I2CFFMProviderImpl.java
+++ b/plugins/pi4j-plugin-ffm/src/main/java/com/pi4j/plugin/ffm/providers/i2c/I2CFFMProviderImpl.java
@@ -29,10 +29,12 @@ public class I2CFFMProviderImpl extends I2CProviderBase implements I2CProvider {
     public I2C create(I2CConfig config) {
         var bus = new I2CBusFFM(config);
 
-        var functions = bus.getFunctionalityMap();
-        logger.debug("{} - bus I2C functions support:", bus.getBusName());
-        for (var entry : functions.entrySet()) {
-            logger.debug("\t{} - {}", entry.getKey(), entry.getValue());
+        if (logger.isDebugEnabled()) {
+            var functions = bus.getFunctionalityMap();
+            logger.debug("{} - bus I2C functions support:", bus.getBusName());
+            for (var entry : functions.entrySet()) {
+                logger.debug("\t{} - {}", entry.getKey(), entry.getValue());
+            }
         }
 
         var impl = config.i2cImplementation();

--- a/plugins/pi4j-plugin-ffm/src/main/java/com/pi4j/plugin/ffm/providers/i2c/I2CFFMProviderImpl.java
+++ b/plugins/pi4j-plugin-ffm/src/main/java/com/pi4j/plugin/ffm/providers/i2c/I2CFFMProviderImpl.java
@@ -5,8 +5,11 @@ import com.pi4j.plugin.ffm.common.PermissionHelper;
 import com.pi4j.plugin.ffm.providers.i2c.impl.I2CDirect;
 import com.pi4j.plugin.ffm.providers.i2c.impl.I2CFile;
 import com.pi4j.plugin.ffm.providers.i2c.impl.I2CSMBus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 public class I2CFFMProviderImpl extends I2CProviderBase implements I2CProvider {
+    private static final Logger logger = LoggerFactory.getLogger(I2CFFMProviderImpl.class);
 
     public I2CFFMProviderImpl() {
         this.id = "ffm-i2c";
@@ -26,16 +29,26 @@ public class I2CFFMProviderImpl extends I2CProviderBase implements I2CProvider {
     public I2C create(I2CConfig config) {
         var bus = new I2CBusFFM(config);
 
+        var functions = bus.getFunctionalityMap();
+        logger.trace("{} - bus I2C functions support:", bus.getBusName());
+        for (var entry : functions.entrySet()) {
+            logger.trace("\t{} - {}", entry.getKey(), entry.getValue());
+        }
+
         var impl = config.i2cImplementation();
         if (impl == null) {
-            impl = I2CImplementation.SMBUS;
+            impl = I2CImplementation.DIRECT;
+            logger.debug("{} - no I2C implementation was chosen, using {} as default", bus.getBusName(), impl);
         }
         I2CBase<?> i2c;
         if (impl.equals(I2CImplementation.SMBUS) && bus.supportsSMBus()) {
+            logger.debug("{} - creating SMBus adapter based on default implementation and functions", bus.getBusName());
             i2c = new I2CSMBus(this, config, bus);
         } else if (impl.equals(I2CImplementation.DIRECT) && bus.supportsDirect()) {
+            logger.debug("{} - creating Direct ioctl adapter based on default implementation and functions", bus.getBusName());
             i2c = new I2CDirect(this, config, bus);
         } else {
+            logger.debug("{} - creating File adapter based on default implementation and functions", bus.getBusName());
             i2c = new I2CFile(this, config, bus);
         }
 

--- a/plugins/pi4j-plugin-ffm/src/main/java/com/pi4j/plugin/ffm/providers/i2c/I2CFFMProviderImpl.java
+++ b/plugins/pi4j-plugin-ffm/src/main/java/com/pi4j/plugin/ffm/providers/i2c/I2CFFMProviderImpl.java
@@ -30,9 +30,9 @@ public class I2CFFMProviderImpl extends I2CProviderBase implements I2CProvider {
         var bus = new I2CBusFFM(config);
 
         var functions = bus.getFunctionalityMap();
-        logger.trace("{} - bus I2C functions support:", bus.getBusName());
+        logger.debug("{} - bus I2C functions support:", bus.getBusName());
         for (var entry : functions.entrySet()) {
-            logger.trace("\t{} - {}", entry.getKey(), entry.getValue());
+            logger.debug("\t{} - {}", entry.getKey(), entry.getValue());
         }
 
         var impl = config.i2cImplementation();

--- a/plugins/pi4j-plugin-ffm/src/test/java/com/pi4j/plugin/ffm/model/I2CTest.java
+++ b/plugins/pi4j-plugin-ffm/src/test/java/com/pi4j/plugin/ffm/model/I2CTest.java
@@ -19,7 +19,7 @@ public class I2CTest {
             var bytes = "Test".getBytes();
             var message = new I2CMessage(1, 2, bytes.length, bytes);
             var buffer = offheap.allocate(I2CMessage.LAYOUT);
-            message.to(buffer);
+            message.to(buffer, offheap);
             var message2 = message.from(buffer);
             assertTrue(new ReflectionEquals(message2).matches(message));
         }
@@ -32,7 +32,7 @@ public class I2CTest {
             var message = new I2CMessage(1, 2, bytes.length, bytes);
             var rdwrData = new RDWRData(new I2CMessage[]{message}, 1);
             var buffer = offheap.allocate(RDWRData.LAYOUT);
-            rdwrData.to(buffer);
+            rdwrData.to(buffer, offheap);
             var rdwrData2 = rdwrData.from(buffer);
             assertEquals(rdwrData.nmsgs(), rdwrData2.nmsgs());
             assertTrue(new ReflectionEquals(rdwrData2.msgs()).matches(rdwrData.msgs()));

